### PR TITLE
sec: Adjust flag validation for TLS.

### PR DESCRIPTION
**Pull request by royjacobson**
This changes our authentication policy when using TLS:

1. If TLS is enabled, some authentication must be used.
2. Password authentication is allowed for client authentication.

### CODE REVIEW

File: src/server/protocol_client.cc [hunk 1]
```diff
@@ -56,17 +56,23 @@ static ProtocolClient::SSL_CTX* CreateSslClientCntx() {
   const auto& tls_key_file = GetFlag(FLAGS_tls_key_file);
   unsigned mask = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
 
-  CHECK_EQ(1, SSL_CTX_use_PrivateKey_file(ctx, tls_key_file.c_str(), SSL_FILETYPE_PEM));
-  const auto& tls_cert_file = GetFlag(FLAGS_tls_cert_file);
+  // Load client certificate if given.
+  if (!tls_key_file.empty()) {
+    CHECK_EQ(1, SSL_CTX_use_PrivateKey_file(ctx, tls_key_file.c_str(), SSL_FILETYPE_PEM));
+    const auto& tls_cert_file = GetFlag(FLAGS_tls_cert_file);
```
Comment (romange) [inline src/server/protocol_client.cc:62]: can `tls_cert_file` be empty in this context?

File: src/server/server_family.cc [hunk 2]
```diff
@@ -419,6 +423,28 @@ void SlowLog(CmdArgList args, ConnectionContext* cntx) {
   (*cntx)->SendError(UnknownSubCmd(sub_cmd, "SLOWLOG"), kSyntaxErrType);
 }
 
+void ValidateServerTlsFlags() {
```
Comment (romange) [inline src/server/server_family.cc:426]: maybe worth adding what this function verifies. That we demand either password or a ca certificate as method for authentication when tls is enabled

File: src/server/protocol_client.cc [hunk 3]
```diff
@@ -56,17 +56,23 @@ static ProtocolClient::SSL_CTX* CreateSslClientCntx() {
   const auto& tls_key_file = GetFlag(FLAGS_tls_key_file);
   unsigned mask = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
 
-  CHECK_EQ(1, SSL_CTX_use_PrivateKey_file(ctx, tls_key_file.c_str(), SSL_FILETYPE_PEM));
-  const auto& tls_cert_file = GetFlag(FLAGS_tls_cert_file);
+  // Load client certificate if given.
+  if (!tls_key_file.empty()) {
```
Comment (kostasrim) [inline src/server/protocol_client.cc:60]: I  am not quite sure what this check adds. From my understanding you want to allow clients(replicas) configured without a private/public key. What's the point of having TLS then? Would this even work?

File: src/server/protocol_client.cc (hunk 1 above)
Comment (kostasrim) [reply src/server/protocol_client.cc:62]: if it is, the later statement will fail :)

File: src/server/protocol_client.cc [hunk 4]
```diff
@@ -56,17 +56,23 @@ static ProtocolClient::SSL_CTX* CreateSslClientCntx() {
...
+  }
 
+  // Load custom certificate validation if given.
   const auto& tls_ca_cert_file = GetFlag(FLAGS_tls_ca_cert_file);
   const auto& tls_ca_cert_dir = GetFlag(FLAGS_tls_ca_cert_dir);
 
   const auto* file = tls_ca_cert_file.empty() ? nullptr : tls_ca_cert_file.data();
   const auto* dir = tls_ca_cert_dir.empty() ? nullptr : tls_ca_cert_dir.data();
-  CHECK_EQ(1, SSL_CTX_load_verify_locations(ctx, file, dir));
+  if (file || dir) {
```
Comment (kostasrim) [inline src/server/protocol_client.cc:73]: CA is  now optional as well? Is that right or am I missing something?

**Review by kostasrim — COMMENTED** [review]
also, plz expand `replication_test.py` with the new flag combinations

File: src/server/protocol_client.cc (hunk 3 above)
Comment (royjacobson) [reply src/server/protocol_client.cc:60]: Yeah, it works (you're correct that I need to add some tests though).
In this configuration the TLS layer provides encryption and server authentication with the server certificate, and the password authenticates the client.

File: src/server/protocol_client.cc (hunk 4 above)
Comment (royjacobson) [reply src/server/protocol_client.cc:73]: Yes, it is optional. AFAIU if it is not given openssl will validate the server certificate against the system default which will include let's encrypt etc. But maybe it misses an additional call to do that, I still have to check with an actual certificate.

File: src/server/protocol_client.cc (hunk 3 above)
Comment (kostasrim) [reply src/server/protocol_client.cc:60]: Ohhh I see 👍

File: src/server/protocol_client.cc (hunk 1 above)
Comment (royjacobson) [reply src/server/protocol_client.cc:62]: no, we do earlier validation in `ValidateClientTlsFlags`. will add a comment

File: src/server/protocol_client.cc (hunk 4 above)
Comment (royjacobson) [reply src/server/protocol_client.cc:73]: it did miss the additional call :laughing: tested manually to verify.

File: src/server/protocol_client.cc [hunk 5]
```diff
@@ -142,22 +151,32 @@ std::string ProtocolClient::ServerContext::Description() const {
   return absl::StrCat(host, ":", port);
 }
 
-void ProtocolClient::ValidateTlsFlags() const {
-  if (absl::GetFlag(FLAGS_tls_cert_file).empty()) {
-    LOG(ERROR) << "tls_cert_file flag should be set";
-    exit(1);
+void ValidateClientTlsFlags() {
```
Comment (kostasrim) [inline src/server/protocol_client.cc:154]:
```suggestion
void ValidateClientTlsFlags() const {
```

File: src/server/protocol_client.cc [hunk 6]
```diff
@@ -142,22 +151,32 @@ std::string ProtocolClient::ServerContext::Description() const {
...
+  }
+
+  bool has_auth = false;
+
+  if (!absl::GetFlag(FLAGS_tls_key_file).empty()) {
+    if (absl::GetFlag(FLAGS_tls_cert_file).empty()) {
+      LOG(ERROR) << "tls_cert_file flag should be set";
+      exit(1);
+    }
+    has_auth = true;
```
Comment (kostasrim) [inline src/server/protocol_client.cc:166]:
```suggestion
    return;
```

File: src/server/protocol_client.cc [hunk 7]
```diff
@@ -142,22 +151,32 @@ std::string ProtocolClient::ServerContext::Description() const {
...
-void ProtocolClient::ValidateTlsFlags() const {
-  if (absl::GetFlag(FLAGS_tls_cert_file).empty()) {
-    LOG(ERROR) << "tls_cert_file flag should be set";
-    exit(1);
+void ValidateClientTlsFlags() {
+  if (!absl::GetFlag(FLAGS_tls_replication)) {
+    return;
+  }
+
+  bool has_auth = false;
```
Comment (kostasrim) [inline src/server/protocol_client.cc:159]:
```suggestion
```

File: src/server/protocol_client.cc [hunk 8]
```diff
@@ -142,22 +151,32 @@ std::string ProtocolClient::ServerContext::Description() const {
...
+      exit(1);
+    }
+    has_auth = true;
   }
 
-  if (absl::GetFlag(FLAGS_tls_ca_cert_file).empty() &&
-      absl::GetFlag(FLAGS_tls_ca_cert_dir).empty()) {
-    LOG(ERROR) << "Either or both tls_ca_cert_file or tls_ca_cert_dir flags must be set";
+  if (!absl::GetFlag(FLAGS_masterauth).empty())
+    has_auth = true;
```
Comment (kostasrim) [inline src/server/protocol_client.cc:170]:
```suggestion
    return;
```

File: src/server/protocol_client.cc [hunk 9]
```diff
@@ -142,22 +151,32 @@ std::string ProtocolClient::ServerContext::Description() const {
...
+
+  if (!has_auth) {
```
Comment (kostasrim) [inline src/server/protocol_client.cc:172]:
```suggestion
```

File: src/server/protocol_client.h [hunk 10]
```diff
@@ -32,6 +32,8 @@ class ConnectionContext;
 class JournalExecutor;
 struct JournalReader;
 
+void ValidateClientTlsFlags();
```
Comment (kostasrim) [inline src/server/protocol_client.h:35]:
```suggestion
void ValidateClientTlsFlags() const;
```

File: src/server/server_family.cc [hunk 11]
```diff
@@ -419,6 +423,32 @@ void SlowLog(CmdArgList args, ConnectionContext* cntx) {
...
+// Check that if TLS is used at least one form of client authentication is
+// enabled. That means either using a password or giving a root
+// certificate for authenticating client certificates which will
+// be required.
+void ValidateServerTlsFlags() {
+  if (!absl::GetFlag(FLAGS_tls)) {
+    return;
+  }
+
+  bool has_auth = false;
```
Comment (kostasrim) [inline src/server/server_family.cc:435]: Same as above, you don't need this boolean, you can just return

File: tests/dragonfly/tls_conf_test.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import pytest
+import redis
+from .utility import *
+from . import DflyStartException
+
```
Comment (kostasrim) [inline tests/dragonfly/tls_conf_test.py:5]: Really nice, now we got tests specifically for TLS :pray:

**Comment by kostasrim** [comment]
So for some reason, I could not mark the whole code block and I could not suggest the changes in one comment. 

Basically, in summary, for the Validate functions you can:

1. Mark them `const` since they don't modify any data member
2. Remove the boolean flag and return. When you set the bool flag to `true` the rest of the checks are redundant

File: src/server/protocol_client.h (hunk 10 above)
Comment (royjacobson) [reply src/server/protocol_client.h:35]: It's a free function :)

**Comment by royjacobson** [comment]
> So for some reason, I could not mark the whole code block and I could not suggest the changes in one comment.
> 
> Basically, in summary, for the Validate functions you can:
> 
>     1. Mark them `const` since they don't modify any data member
> 
>     2. Remove the boolean flag and return. When you set the bool flag to `true` the rest of the checks are redundant

I disagree about the `has_auth` changes - I wrote this code in this way so it's easy to understand exactly what properties the function guarantees and to validate it works correctly. I feel that early returns make it a bit less clear (even though it's longer).

About your first point, those are free functions :)

File: tests/dragonfly/tls_conf_test.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (kostasrim) [inline tests/dragonfly/tls_conf_test.py:6]: Can you also try to connect with a client on all the test cases?

**Comment by royjacobson** [comment]
@kostasrim 1 - some combinations of flags (like 'key' flag without a 'cert' flag) result in exception instead of an early return. And 2, I agree about early return in general but I feel this is a better way to write security critical code like auth policies :sweat_smile:. So I'd like to stick with this version

**Review by kostasrim — APPROVED** [review]

**Pull request merged**